internal/exec/validate.go

```diff
@@ -0,0 +1,108 @@
+package exec
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/graph-gophers/graphql-go/errors"
+	"github.com/graph-gophers/graphql-go/internal/exec/resolvable"
+	"github.com/graph-gophers/graphql-go/internal/exec/selected"
+	"github.com/graph-gophers/graphql-go/types"
+)
+
+func collectFieldsToValidate(sels []selected.Selection, s *resolvable.Schema, fields *[]*fieldToValidate, fieldByAlias map[string]*fieldToValidate) {
+	for _, sel := range sels {
+		switch sel := sel.(type) {
+		case *selected.SchemaField:
+			field, ok := fieldByAlias[sel.Alias]
+			if !ok { // validation already checked for conflict (TODO)
+				field = &fieldToValidate{field: sel}
+				fieldByAlias[sel.Alias] = field
+				*fields = append(*fields, field)
+			}
+			field.sels = append(field.sels, sel.Sels...)
+		case *selected.TypenameField:
+			// TODO: confirm behaviour for this case
+			// Ignore __typename, which has no directives
+		case *selected.TypeAssertion:
+			collectFieldsToValidate(sel.Sels, s, fields, fieldByAlias)
+		default:
+			panic(fmt.Sprintf("unexpected selection type %T", sel))
+		}
+	}
+}
+
+func validateFieldSelection(ctx context.Context, s *resolvable.Schema, f *fieldToValidate, path *pathSegment) []*errors.QueryError {
+	if f.field.FixedResult.IsValid() {
+		// Skip fixed result meta fields like __TypeName
+		return nil
+	}
+
+	var args interface{}
+
+	if f.field.ArgsPacker != nil {
+		args = f.field.PackedArgs.Interface()
+	}
+
+	vErrs := f.field.Validate(ctx, args)
+
+	if l := len(vErrs); l > 0 {
+		errs := make([]*errors.QueryError, l)
+
+		for i, err := range vErrs {
+			var ext map[string]interface{}
+			if e, ok := err.(extensionser); ok {
+				ext = e.Extensions()
+			}
+			errs[i] = &errors.QueryError{
+				Err:        err,
+				Message:    err.Error(),
+				Locations:  []errors.Location{f.field.Loc}, // TODO: is this the correct location?
```

Comment:
LGTM

Reply:
I should’ve added more context in the comment here: I’m unclear whether the location should be where the field is defined in the _schema definition_, or where in the _requested query_ the field is

Reply:
We are validating the query.

Reply:
This is the location in the schema being applied here, rather than the query. I'm not sure that we have access to the other